server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -259,16 +282,31 @@ private static List<MoveAllocationCommand> getMoveCommands(AllocationCommands co
     }
 
     private void setCurrentDesiredBalance(DesiredBalance newDesiredBalance) {
-        if (logger.isTraceEnabled()) {
-            var diff = DesiredBalance.hasChanges(currentDesiredBalance, newDesiredBalance)
-                ? "Diff: " + DesiredBalance.humanReadableDiff(currentDesiredBalance, newDesiredBalance)
-                : "No changes";
-            logger.trace("Desired balance updated: {}. {}", newDesiredBalance, diff);
+        // Update current desired balance if the old value has not been changed already by master fail-over
+        final AtomicReference<DesiredBalance> oldDesiredBalance = new AtomicReference<>();
+        final DesiredBalance updatedDesiredBalance = currentDesiredBalanceRef.updateAndGet(current -> {
+            if (current != DesiredBalance.NOT_MASTER) {
+                oldDesiredBalance.set(current);
+                return newDesiredBalance;
+            } else {
+                return current;
+            }
+        });
+
+        if (updatedDesiredBalance == newDesiredBalance) {
+            assert oldDesiredBalance.get() != null;
+            if (logger.isTraceEnabled()) {
+                var diff = DesiredBalance.hasChanges(oldDesiredBalance.get(), newDesiredBalance)
+                    ? "Diff: " + DesiredBalance.humanReadableDiff(oldDesiredBalance.get(), newDesiredBalance)
+                    : "No changes";
+                logger.trace("Desired balance updated: {}. {}", newDesiredBalance, diff);
+            } else {
+                logger.debug("Desired balance updated for [{}]", newDesiredBalance.lastConvergedIndex());
+            }
+            computedShardMovements.inc(DesiredBalance.shardMovements(oldDesiredBalance.get(), newDesiredBalance));
```

Comment:
Alternative approach? not sure if it's better or worse...

```java
while (true) {
    final DesiredBalance previousBalance = currentDesiredBalanceRef.get();
    if (previousDesiredBalance == DesiredBalance.NOT_MASTER) {
        logger.debug("discard desired balance for [{}]", newDesiredBalance.lastConvergedIndex());
        break;
    }
    if (currentDesiredBalanceRef.get().compareAndSet(previousDesiredBalance, newDesiredBalance) {
        if (logger.isTraceEnabled()) {
            var diff = DesiredBalance.hasChanges(previousDesiredBalance, newDesiredBalance)
                ? "Diff: " + DesiredBalance.humanReadableDiff(previousDesiredBalance, newDesiredBalance)
                : "No changes";
            logger.trace("Desired balance updated: {}. {}", newDesiredBalance, diff);
        } else {
            logger.debug("Desired balance updated for [{}]", newDesiredBalance.lastConvergedIndex());
        }
        computedShardMovements.inc(DesiredBalance.shardMovements(previousDesiredBalance, newDesiredBalance));
        break;
    }
}
```

Reply:
I think that's kind-of what `updateAndGet` does internally, because the javadoc says:

> The function should be side-effect-free, since it may be re-applied when attempted updates fail due to contention among threads.

Reply:
That's viable as well. I don't have strong opinions. The concurrency here should be fairly low since it happens only during master failover.

Reply:
> I think that's kind-of what updateAndGet does internally, because the javadoc says:

I think so too. We also use this pattern elsewhere in code.

I pushed b5cd708b60fb855120985d752c034e1599ef4647 to apply your suggeston.

I think it has an edge-case advantage to return earlier when the 1st read from `currentDesiredBalanceRef` is a not-master. The `updateAndGet` would probably try to set it again which may fail and trigger retry and see a different value which is from a newer term.